snap/sysparams/sysparams.go

```diff
@@ -0,0 +1,109 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package sysparams
+
+import (
+	"bufio"
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	"github.com/snapcore/snapd/osutil"
+)
+
+// For mocking in tests
+var (
+	osutilAtomicWriteFile = osutil.AtomicWriteFile
+)
+
+// SystemParams manages certain system configuration parameters
+// aspects like the homedirs configuration that must be available
+// for other binaries, as snap-confine.
+type SystemParams struct {
+	// path stored to allow for updating the same path
+	path string
+	// Homedirs is the comma-delimited list of user specified home
+	// directories that should be mounted
+	Homedirs string
+}
+
+func parseSystemParams(contents string) (*SystemParams, error) {
+	params := &SystemParams{}
+	seen := make(map[string]bool)
+	scanner := bufio.NewScanner(strings.NewReader(contents))
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		// ignore empty lines and comments
+		if len(line) == 0 || strings.HasPrefix(line, "#") {
+			continue
+		}
+
+		tokens := strings.SplitN(line, "=", 2)
+		if len(tokens) != 2 {
+			return nil, fmt.Errorf("cannot parse invalid line: %q", line)
+		}
+
+		// ensure that each configuration value only appears once
+		if ok := seen[tokens[0]]; ok {
```

Comment:
I'm not familiar with the system parameters file and its usage, can it have other entries besides "homedirs" (ignoring comments and empty lines)? Or, if not, is this more generic for forwards compatibility in case we ever other types of entries?

Reply:
Good question, and I'd say both forward-compatability situation, and of neccessity. There was no other natural place for the homedirs stuff to go, and it something that needs to be shared with snap-confine (snap-confine will load this file as well).

Reply:
yes, the idea was to introduce a general mechanism fro snapd to communicate parameters to snap-confine or similar other processes that cannot reasonably access state directly

Reply:
If "homedirs" was the only config we could have then the logic to check for double entries could be simplified to just setting/checking a flag but if it's for forwards compatibility in case we add new entries then that makes sense. Thanks for the reply